<?xml version="1.0" encoding="utf-8"?>
<launch>

<param name="/use_sim_time" value="true"/>

### FLOORPLAN RECONSTRUCTION ###

  <node pkg="sigmafp" type="fake_amcl.py" name="fake_amcl" launch-prefix="">
      <param name="mode" value="tf"/>
      <param name="topic_name" value="none"/>
      <param name="world_name" value="world"/>
      <param name="baselink_name" value="base_link_gt"/>
     </node>

  <node pkg="sigmafp" type="tf_br.py" name="tf_br" launch-prefix="">
  </node>

  <node name="sigmafp" pkg="sigmafp" type="sigma-FP.py" output="screen">

      ### Dataset info ###
      <param name="dataset" value="uHumans2"/>
      <param name="data_mode" value="gt"/>

      ### Topics names ###
      <param name="topic_cameraRGB" value="tesse/left_cam/rgb/image_raw"/>
      <param name="topic_cameraDepth" value="tesse/depth_cam/mono/image_raw"/>
      <param name="topic_result" value="ViMantic/Detections"/>
      <param name="topic_republic" value="ViMantic/ToCNN"/>
      <param name="topic_cnn" value="/tesse/seg_cam/rgb/image_raw"/>
      <param name="topic_localization" value="amcl_pose"/>

      ### Camera calibration ###
      <param name="image_width" value="720"/>
      <param name="image_height" value="480"/>
      <param name="camera_cx" value="360"/>
      <param name="camera_cy" value="240"/>
      <param name="camera_fx" value="415.69219381653056"/>
      <param name="camera_fy" value="415.69219381653056"/>
      <param name="camera_depth_max_range" value="10.0"/>

      ### SigmaFP parameters ###
      <param name="points_in_pcd" value="4000"/>
      <param name="min_points_plane" value="100"/>
      <param name="min_plane_width" value="0.1"/> # In m
      <param name="min_px_opening" value="1000"/>
      <param name="min_area_opening" value="1000"/>  # In cm2
      <param name="bhattacharyya_threshold" value="1."/>
      <param name="euclidean_threshold" value="0.3"/>
      <param name="eps_alpha" value="8.0"/> # 10
      <param name="eps_beta" value="4.0"/> # 5
      <param name="eps_dist" value="0.02"/>

      ### SigmaFP-textures parameters ###
      <param name="extract_textures" value="false"/>
      <param name="feature_type" value="ORB"/>

      ### Saving data ###
      <param name="save_colmap" value="true"/>
      <param name="data_category" value="localization"/>
      <param name="save_path" value="/home/eostajm/datasets/uhumans2/results/"/>
      <param name="map_path" value="/home/eostajm/datasets/uhumans2/results/wallmap_no_refined.npy"/>
      
      ### Flags ###
      <param name="debug" value="true"/>
      <param name="mesh_to_unity" value="false"/>
      
  </node>

    ### NAVIGATION ###
    <arg name="map_file" value="/home/eostajm/datasets/uhumans2/maps/uHumans2_office_s1_00h.yaml"/>
    <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" launch-prefix="xterm -e">
      <param name="frame_id" value="map"/>  
    </node>

    ### RVIZ ###
    <node name="rviz_player" pkg="rviz" type="rviz"
          args="-d $(find sigmafp)/rviz/uhumans2.rviz"/>

     ### ROBOT LOCALIZATION ###
     
    <!-- 
      
      <remap from="scan" to="/tesse/front_lidar/scan"/>  topic id to read scan lasers from 
      <arg name="launch_prefix" default="" />
      <node pkg="amcl" type="amcl" name="amcl" launch-prefix="xterm -e" output="log">
        ### Initial pose
        <param name="initial_pose_x" value="0.0"/>         #m
        <param name="initial_pose_y" value="0.0"/>        #m
        <param name="initial_pose_a" value="0.0"/>     #rad

        ### Filter params
        <param name="min_particles" value="100"/>   #200
        <param name="max_particles" value="1000"/>
        <param name="kld_err" value="0.05"/>
        <param name="kld_z" value="0.99"/>
        <param name="update_min_d" value="0.0"/>    ### Translational movement required before performing a filter update.
        <param name="update_min_a" value="0.0"/>    ### Rotational movement required before performing a filter update. (rad)
        <param name="resample_interval" value="2"/>
        <param name="transform_tolerance" value="0.2"/> #0.2
        <param name="recovery_alpha_slow" value="0.001"/> #0.001
        <param name="recovery_alpha_fast" value="0.1"/> #0.1
        <param name="gui_publish_rate" value="-1.0"/>   ### 10 Maximum rate (Hz) at which scans and paths are published for visualization, -1.0 to disable.

        ### Laser model
        <param name="laser_model_type" value="likelihood_field"/>   ###The likelihood_field model uses only: z_hit and z_rand.
        <param name="laser_max_beams" value="120"/> #30
        <param name="laser_z_hit" value="0.5"/>
        <param name="laser_z_rand" value="0.5"/>
        <param name="laser_z_short" value="0.05"/>
        <param name="laser_z_max" value="0.05"/>

        ### Odometry model
        <param name="odom_model_type" value="omni"/>
        <param name="odom_alpha1" value="0.5"/>     ### expected noise in odometry's rotation estimate from the rotational component
        <param name="odom_alpha2" value="0.5"/>     ### expected noise in odometry's rotation estimate from the translational component
        <param name="odom_alpha3" value="0.5"/>     ### expected noise in odometry's translation estimate from the translational component
        <param name="odom_alpha4" value="0.5"/>     ### expected noise in odometry's translation estimate from the rotational component
        <param name="odom_alpha5" value="0.5"/>     ### Translation-related noise parameter (only used if model is "omni").
        <param name="odom_frame_id" value="odom"/>
        <param name="base_frame_id" value="base_link_gt"/>
        <param name="global_frame_id" value="map"/>
        <param name="tf_broadcast" value="true"/>
      </node>
      
      ### ODOMETRY RF2O ###
    <node pkg="rf2o_laser_odometry" type="rf2o_laser_odometry_node" name="rf2o_laser_odometry" launch-prefix="xterm -e" output="log">
        <param name="laser_scan_topic" value="/tesse/front_lidar/scan"/>        ### topic where the lidar scans are being published
        <param name="odom_topic" value="/odom" />                   ### topic where to publish the odometry estimations
        <param name="publish_tf" value="true" />                    ### wheter or not to publish the tf::transform (base->odom)
        <param name="base_frame_id" value="/base_link_gt"/>            ### frame_id (tf) of the mobile robot base. A tf transform from the laser_frame to the base_frame is mandatory
        <param name="odom_frame_id" value="/odom" />                ### frame_id (tf) to publish the odometry estimations
        <param name="freq" value="30.0"/>                            ### Execution frequency. See "Planar Odometry from a Radial Laser Scanner. A Range Flow-based Approach. ICRA'16"
        <param name="init_pose_from_topic" value="" />
     </node>
      -->

</launch>